# Power Predictor - 07-08-2024

Hello, world! In this project I put to use my knowledge of Linear Regression to build a model that predicts the power consumption for a given household based on a few factors(predictors).

I will start by trying to understand my dataset. Lets explore!

In [1]:
import pandas as pd

acorn_details = pd.read_csv("Powerdata/acorn_details.csv", encoding='utf-8', encoding_errors='replace')
#Pandas could not read some of the data in the file due to some encoding error, hence i have had to include the 'encoding'
#parameters withion the 'read_csv' method
acorn_details.tail(20)

MAIN CATEGORIES                       CATEGORIES  \
806    LEISURE TIME      Visit Pubs for a Meal - Day   
807    LEISURE TIME      Visit Pubs for a Meal - Day   
808    LEISURE TIME  Visit Pubs for a Meal - Evening   
809    LEISURE TIME  Visit Pubs for a Meal - Evening   
810    LEISURE TIME  Visit Pubs for a Meal - Evening   
811    LEISURE TIME         Restaurants - Most Often   
812    LEISURE TIME         Restaurants - Most Often   
813    LEISURE TIME         Restaurants - Most Often   
814    LEISURE TIME         Holiday Destination/Type   
815    LEISURE TIME         Holiday Destination/Type   
816    LEISURE TIME         Holiday Destination/Type   
817    LEISURE TIME         Holiday Destination/Type   
818    LEISURE TIME         Holiday Destination/Type   
819    LEISURE TIME         Holiday Destination/Type   
820    LEISURE TIME         Holiday Destination/Type   
821    LEISURE TIME         Holiday Destination/Type   
822    LEISURE TIME         Holiday Destination/Type   
823    LEISURE TIME         Holiday Destination/Type   
824    LEISURE TIME         Holiday Destination/Type   
825    LEISURE TIME         Holiday Destination/Type   

                     REFERENCE  ACORN-A  ACORN-B  ACORN-C  ACORN-D  ACORN-E  \
806                Mass Market    103.0    107.0    103.0    137.0    137.0   
807                      Value     64.0     97.0     96.0     93.0     95.0   
808                    Premium     95.0    114.0     79.0    308.0    218.0   
809                Mass Market    100.0    115.0     97.0    204.0    202.0   
810                      Value     97.0    100.0    100.0    111.0    111.0   
811                    Premium    184.0    174.0    123.0    224.0    163.0   
812                Mass Market    163.0    136.0     99.0    150.0    150.0   
813                      Value     65.0     95.0     94.0     45.0     78.0   
814               UK / Ireland    116.0    117.0    116.0    101.0    101.0   
815      Spain / Mediterranean    103.0    103.0    103.0    103.0    103.0   
816             Europe (other)    402.0    149.0     85.0    214.0    171.0   
817               USA / Canada   2280.0    344.0     79.0     84.0     56.0   
818                  Caribbean    133.0    150.0    150.0    153.0    128.0   
819    Australia / New Zealand   3795.0    116.0    116.0     58.0     58.0   
820                     Africa   2652.0    126.0     88.0    424.0    108.0   
821                       Asia    171.0    137.0     94.0    220.0    196.0   
822  Activity / Outdoor Sports    298.0    278.0    138.0    119.0     93.0   
823                     Cruise    272.0    295.0    272.0     44.0     44.0   
824                    Package    196.0    186.0    166.0     49.0     49.0   
825              Self-catering    308.0    181.0    181.0     55.0     55.0   

     ACORN-F  ACORN-G  ACORN-H  ACORN-I  ACORN-J  ACORN-K  ACORN-L  ACORN-M  \
806     50.0    103.0    101.0     76.0    145.0    145.0     91.0     86.0   
807    100.0     99.0    100.0    100.0    100.0    123.0    101.0    100.0   
808     73.0     90.0     90.0     75.0    100.0    149.0     78.0     75.0   
809     56.0     92.0     92.0     65.0    127.0    183.0     76.0     70.0   
810     81.0    100.0    113.0     72.0    111.0    165.0     99.0     99.0   
811     98.0    116.0    101.0    101.0    118.0     62.0     62.0     62.0   
812     74.0    116.0    105.0     76.0    119.0    141.0     79.0     69.0   
813     82.0    108.0    128.0    111.0    108.0     79.0    105.0    117.0   
814    119.0    110.0    109.0    113.0    110.0     77.0     90.0     87.0   
815     80.0     98.0     98.0     78.0     98.0    179.0    101.0    103.0   
816     97.0     97.0     61.0     97.0     97.0    318.0     67.0     45.0   
817     62.0     62.0     51.0     62.0     62.0     30.0     30.0     30.0   
818     81.0    137.0    114.0    112.0    112.0     65.0     62.0     65.0   
819     58.0     58.0     58.0     58.0     58.0     58.0    

This file contains a summary describing the people who stay in different acorns. It paints a picture as to what each acorn looks like and the kind of people who dwell there. This is important for determining electricity consumption. I will have to figure out how to leverage these data later on. Forn now, we explore another file from our data set.

In [2]:
infor_households = pd.read_csv("Powerdata/informations_households.csv")
infor_households.head(500)

LCLid stdorToU    Acorn Acorn_grouped     file
0    MAC005492      ToU   ACORN-        ACORN-  block_0
1    MAC001074      ToU   ACORN-        ACORN-  block_0
2    MAC000002      Std  ACORN-A      Affluent  block_0
3    MAC003613      Std  ACORN-A      Affluent  block_0
4    MAC003597      Std  ACORN-A      Affluent  block_0
..         ...      ...      ...           ...      ...
495  MAC002380      Std  ACORN-D      Affluent  block_9
496  MAC004760      Std  ACORN-D      Affluent  block_9
497  MAC000026      Std  ACORN-D      Affluent  block_9
498  MAC002005      ToU  ACORN-D      Affluent  block_9
499  MAC001613      Std  ACORN-D      Affluent  block_9

[500 rows x 5 columns]

So, I am also simultaneously trying to understand the structure of the UK society with regrads to the information contained in this data set. I have been reading online as well as the notes from the Author on Kaggle as well as the discussions surrounding this dataset and I must say that, that is making the dataset easier to understand.
The file above basicall lists all the houses involved, each with a unique id(LCid), the type of tarriffs that apply to them as well as the acorn group the households are in and the file within the dataset where the household meter data is found.

In [3]:
daily_energy = pd.read_csv("Powerdata/daily_dataset/daily_dataset/block_0.csv")
daily_energy.head(60)

LCLid         day  energy_median  energy_mean  energy_max  \
0   MAC000002  2012-10-12         0.1385     0.154304       0.886   
1   MAC000002  2012-10-13         0.1800     0.230979       0.933   
2   MAC000002  2012-10-14         0.1580     0.275479       1.085   
3   MAC000002  2012-10-15         0.1310     0.213688       1.164   
4   MAC000002  2012-10-16         0.1450     0.203521       0.991   
5   MAC000002  2012-10-17         0.1480     0.226771       0.784   
6   MAC000002  2012-10-18         0.1660     0.223979       0.897   
7   MAC000002  2012-10-19         0.1340     0.175646       0.822   
8   MAC000002  2012-10-20         0.2310     0.362042       1.453   
9   MAC000002  2012-10-21         0.3850     0.510208       2.160   
10  MAC000002  2012-10-22         0.3935     0.393437       1.126   
11  MAC000002  2012-10-23         0.1615     0.218437       0.671   
12  MAC000002  2012-10-24         0.1860     0.323688       1.889   
13  MAC000002  2012-10-25         0.1760     0.273500       0.917   
14  MAC000002  2012-10-26         0.1885     0.313854       0.962   
15  MAC000002  2012-10-27         0.3190     0.351792       1.228   
16  MAC000002  2012-10-28         0.3130     0.408937       1.173   
17  MAC000002  2012-10-29         0.1755     0.266229       1.058   
18  MAC000002  2012-10-30         0.2275     0.290854       0.893   
19  MAC000002  2012-10-31         0.2685     0.371292       1.445   
20  MAC000002  2012-11-01         0.2050     0.254354       0.964   
21  MAC000002  2012-11-02         0.2600     0.309660       1.284   
22  MAC000002  2012-11-03         0.2665     0.298896       1.137   
23  MAC000002  2012-11-04         0.2675     0.302229       1.111   
24  MAC000002  2012-11-05         0.2025     0.314479       1.078   
25  MAC000002  2012-11-06         0.1790     0.242979       1.156   
26  MAC000002  2012-11-07         0.1940     0.273687       1.219   
27  MAC000002  2012-11-08         0.1860     0.186000       0.186   
28  MAC000002  2012-11-09         0.2430     0.355787       1.714   
29  MAC000002  2012-11-10         0.2160     0.275938       1.062   
30  MAC000002  2012-11-11         0.1850     0.222896       1.149   
31  MAC000002  2012-11-12         0.2050     0.256688       1.124   
32  MAC000002  2012-11-13         0.1730     0.213833       0.907   
33  MAC000002  2012-11-14         0.1745     0.225417       1.154   
34  MAC000002  2012-11-15         0.1540     0.204333       0.962   
35  MAC000002  2012-11-16         0.2170     0.273250       1.341   
36  MAC000002  2012-11-17         0.1620     0.203542       0.806   
37  MAC000002  2012-11-18         0.2310     0.287062       1.207   
38  MAC000002  2012-11-19         0.1360     0.211396       0.796   
39  MAC000002  2012-11-20         0.1980     0.233771       1.188   
40  MAC000002  2012-11-21         0.1965     0.221604       1.174   
41  MAC000002  2012-11-22         0.1405     0.196667       0.971   
42  MAC000002  2012-11-23         0.1565     0.209396       1.011   
43  MAC000002  2012-11-24         0.1945     0.229646       1.086   
44  MAC000002  2012-11-25         0.1930     0.219688       0.917   
45  MAC000002  2012-11-26         0.1360     0.205750       1.097   
46  MAC000002  2012-11-27         0.1935     0.228708       1.020   
47  MAC000002  2012-11-28         0.1975     0.223812       1.192   
48  MAC000002  2012-11-29         0.1515     0.224396       1.122   
49  MAC000002  2012-11-30         0.2210     0.272667       0.919   
50  MAC000002  2012-12-01         0.2160     0.266312       1.156   
51  MAC000002  2012-12-02         0.2395     0.319875       1.327   
52  MAC000002  2012-12-03         0.1340     0.211812       1.027   
53  MAC000002  2012-12-04         0.1625     0.232146       0.966   
54  MAC000002  2012-12-05         0.1535     0.244042       1.066   
55  MAC000002  2012-12-06         0.1435     0.246875       0.983   
56  MAC000002  2012-12-07         0.1930     0.276000       1.124   
57  MA

This file and the 'half_hourly' one, provide the meter readings for each household on a daily and half-hourly basis respectively. Overall, gifferent households are found in different blocks and different blocks are found in different acorn groups (which are based on the different social and economic statutes of the households).

For my application, I think I would like to make use of daily records, instaed of hourly records and the 'maximum energy consumption for a household' to be precise. This approach ensures that I include all the meter data for the day. Hourly reading can be useful, I mean they would provide a much more broadly useful model e.g. to predict energy consumption hourly or even half hourly. I might consider that as  a future project but for now, daily records seem much more meaniful for me. This also means less predictors/variables are used resulting in a less sophisticated model.

In that regard, I will use the 'daily datasets' as starting point for my analysis. The datasets, are quite too  many and huge. That could a good sign; the more the data, the better the model. But also, it is going to be a lot of data crunching, let's find out:

In [4]:
df = pd.read_csv("Powerdata/daily_dataset/daily_dataset/block_0.csv")
df

LCLid         day  energy_median  energy_mean  energy_max  \
0      MAC000002  2012-10-12         0.1385     0.154304       0.886   
1      MAC000002  2012-10-13         0.1800     0.230979       0.933   
2      MAC000002  2012-10-14         0.1580     0.275479       1.085   
3      MAC000002  2012-10-15         0.1310     0.213688       1.164   
4      MAC000002  2012-10-16         0.1450     0.203521       0.991   
...          ...         ...            ...          ...         ...   
25569  MAC005492  2014-02-24         0.1690     0.175042       0.378   
25570  MAC005492  2014-02-25         0.1550     0.160792       0.545   
25571  MAC005492  2014-02-26         0.1490     0.178542       0.687   
25572  MAC005492  2014-02-27         0.1140     0.146167       0.478   
25573  MAC005492  2014-02-28         0.0880     0.088000       0.088   

       energy_count  energy_std  energy_sum  energy_min  
0                46    0.196034       7.098       0.000  
1                48    0.192329      11.087       0.076  
2                48    0.274647      13.223       0.070  
3                48    0.224483      10.257       0.070  
4                48    0.184115       9.769       0.087  
...             ...         ...         ...         ...  
25569            48    0.073174       8.402       0.079  
25570            48    0.082118       7.718       0.079  
25571            48    0.120820       8.570       0.079  
25572            48    0.082616       7.016       0.079  
25573             1         NaN       0.088       0.088  

[25574 rows x 9 columns]

In [5]:
#Only going to make use of the maximum _daily-energy consumption per household 'energy_max'
energy_df = df.drop(['energy_count','energy_median','energy_mean','energy_std','energy_sum','energy_min'],axis=1)
energy_df

LCLid         day  energy_max
0      MAC000002  2012-10-12       0.886
1      MAC000002  2012-10-13       0.933
2      MAC000002  2012-10-14       1.085
3      MAC000002  2012-10-15       1.164
4      MAC000002  2012-10-16       0.991
...          ...         ...         ...
25569  MAC005492  2014-02-24       0.378
25570  MAC005492  2014-02-25       0.545
25571  MAC005492  2014-02-26       0.687
25572  MAC005492  2014-02-27       0.478
25573  MAC005492  2014-02-28       0.088

[25574 rows x 3 columns]

Now I have only data, important for my project. But this is only one block out of 111 blocks of households!😓

In [6]:
# To add all the other 110 blocks:
for i in range(1,112):
    temp_df = pd.read_csv(f"Powerdata/daily_dataset/daily_dataset/block_{i}.csv")
    #Delete all the other columns only keeep the useful ones:
    temp_df = temp_df.drop(['energy_count','energy_median','energy_mean','energy_std','energy_sum','energy_min'],axis=1)
    #Then stack/merge the data to the energy dataframe definded above:
    energy_df = pd.concat([energy_df,temp_df],axis=0)
energy_df

LCLid         day  energy_max
0     MAC000002  2012-10-12       0.886
1     MAC000002  2012-10-13       0.933
2     MAC000002  2012-10-14       1.085
3     MAC000002  2012-10-15       1.164
4     MAC000002  2012-10-16       0.991
...         ...         ...         ...
9903  MAC004828  2014-02-24       0.722
9904  MAC004828  2014-02-25       0.051
9905  MAC004828  2014-02-26       0.152
9906  MAC004828  2014-02-27       0.092
9907  MAC004828  2014-02-28       0.000

[3510433 rows x 3 columns]

Now I have all the households in this dataset with their maximum consumptions per day. This is the biggest dataset I have ever worked with and I must say that the size alone, is making nervous and excited at the same time😂

In [7]:
#I will quickly check if there are any missing values so far:
energy_df.isna().sum()

LCLid          0
day            0
energy_max    30
dtype: int64

So, I have got **60** missing values in my dataframe, in the energy_max column. I will have to check them later when I have fully composed my dataframe, so that I can be better informed on what to do with them.


At this point, it might be useful to pose the questions that i intend to answer from this data. These questions will also determine my next step in this project, ***Feature Selection***. I cannot build a meaningful predictive model If I am not clear what factors(predictors) affect power consumption. 

**Here are some of the questions:**

* _Is there a relationship between any of the factors presented in this data; weather, economic status/ social status of a household or whetehr or not a day is a holiday, with the power consumption for that particular day?_

* _How strong is the relationship between the power consumption of a household and each of those factors(if ther ehappen to be a realationship)? If we don't put a particular factor into consideration, would it affect the power consumption for a household?_

* _Can we predict the power consumption of a household based on these data. If so, is linear regression the appropriate model to use to make those predictions?_

There, are some of the questions we could probe from this data and of course a whole lot of inference can also be done by adequately analysing this dataset. 

Now, let's dive right into it!

In [8]:

energy_df = energy_df.rename(columns={'LCLid':'Household_id'})#making the household id column more meaningful
#And setting the household id as the index.
energy_df = energy_df.astype({
    'Household_id': 'str',
    'day': 'object',
    'energy_max': 'float64',
}, errors='raise')
energy_df = energy_df.set_index('Household_id')



In [9]:
weather_df = pd.read_csv('Powerdata/informations_households.csv')
weather_df = weather_df.rename(columns={'LCLid':'Household_id'}).drop(['file'],axis=1)


# Convert the data types of the relevant columns
weather_df = weather_df.astype({
    'Household_id': 'str',
    'Acorn': 'str',
    'stdorToU': 'str',
    'Acorn_grouped': 'str'
}, errors='raise')


weather_df = weather_df.set_index('Household_id')
# Check the data types
print(weather_df.dtypes)
weather_df.columns
print(weather_df[['Acorn', 'stdorToU', 'Acorn_grouped']].isnull().sum())


stdorToU         object
Acorn            object
Acorn_grouped    object
dtype: object
Acorn            0
stdorToU         0
Acorn_grouped    0
dtype: int64


I need to merge the two dataframes i.e. to paste the columns of the previous dataframe onto my energe_df.

In [10]:
# Helper function to match index and assign value
def add_col(energy_df, weather_df, col):
    # Create a new column with default empty strings
    energy_df[col] = ''
    
    # Get the list of indexes
    energy_indexes = energy_df.index.tolist()
    weather_indexes = weather_df.index.tolist()
    
    
    # Loop over the index list
    for row in energy_indexes:
        if row in weather_indexes:
            energy_df.at[row, col] = weather_df.loc[row, col]

    return energy_df



In [11]:
energy_df = add_col(energy_df, weather_df,'Acorn')

This function really takes long(almost 3 minutes) to execute. I might have to look for ways to optimiize it. For now, I will focus on execution.

_Actually, it was taking about 3 minutes when the add_col function was mulfunctioning. I was getting an empty new column after 3 minutes of the function running. Now it has been running for over 30 minutes. I am guessing that that is a good sign, considering the giant size of my dataframe...over 3 million rows in the energy_df. However, I think I still have much to learn, I must look into multiprocessing and batch processing techniques among other optimization tools. For now I wait to see whta is going to happen after many minutes or running the code cell above!_

In [ ]:
#energy_df = add_col(energy_df, weather_df,'stdorToU')

In [ ]:
#energy_df = add_col(energy_df, weather_df,'Acorn_grouped')

In [ ]:
energy_df

In [ ]:
energy_df.dtypes